deploy/init-db/db.sql.txt

```diff
@@ -0,0 +1,62 @@
+-- 'resources' is used to store all of our resources. See comments below for an explanation of the columns.
+CREATE TABLE resources (
+    -- example: "/planes/radius/local/resourceGroups/rg1/providers/Applications.Core/applications/my-app"
+    --
+    -- We use columns to break out the most important components of the resource id for optimal querying.
+    --
+    -- Since resource ids are case-insensitive we canonicalize these columns to lowercase.
+    -- We store the original resource id with the original casing so users can work with their preferred
+    -- naming/casing conventions.
+    --
+    -- We ensure a leading and trailing slash on the components of the resource id for ease of comparison.
+    --
+    -- id -> "/planes/radius/local/resourcegroups/rg1/providers/applications.core/applications/my-app/"
+    -- resource_type -> "/applications.core/applications/"
+    -- root_scope -> "/planes/radius/local/resourcegroups/rg1/"
+    -- routing_scope -> "/applications.core/applications/my-app/"
+
+    -- resource id used as key.
+    id TEXT PRIMARY KEY NOT NULL,
+
+    -- original_id is used to store the original id of the resource before any normalization occurs.
+    -- This is provided for compatability with the existing design of the store API, and can be removed
+    -- in the future.
+    original_id TEXT NOT NULL,
+
+    -- resource type by queries to filter by type.
+    resource_type TEXT NOT NULL,
+
+    -- root_scope used by queries to list resources by their scope.
+    root_scope TEXT NOT NULL,
+
+    -- routing_scope used by queries to list resources when they are child resources.
+    routing_scope TEXT NOT NULL,
+    
+    -- etag used for optimistic concurrency control.
+    etag TEXT NOT NULL,
+
+    -- timestamp is used to implement cursor-based pagination (see below).
+    created_at TIMESTAMP (6) WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
+
+    -- resource_data stores the resource data.
+    resource_data JSONB NOT NULL
+);
+
+-- idx_resource_query is an index for improving performance of queries.
+--
+-- Queries always list resources by their:
+--  - resource_type, and root_scope OR
+--  - resource_type, root_scope, and routing_scope
+--
+-- eg: "find all Applications.Core/applications resources in /planes/radius/local/resourceGroups/my-rg"
+--
+-- > "resource_type" = "/applications.core/applications/"
+-- > "root_scope" = "/planes/radius/local/resourcegroups/my-rg"
+-- > "routing_scope" = NULL
+--
+-- 'created_at' is used with ORDER BY to sort the output, so we can implement cursor-based pagination.
+--
+-- 1) For the initial query, we won't specify a cursor value.
+-- 2) For the next query, we will specify the cursor value as the last created_at value from the previous
+--    query, which allows us to skip the records that were already returned.
+CREATE INDEX idx_resource_query ON resources (resource_type, root_scope, routing_scope, created_at);
```

Comment:
Should we create separate indexes based on the query patterns? Based on what you said above, I believe having two separate indexes might be better:
- `resource_type, and root_scope` => INDEX (resource_type, root_scope, created_at)
- `resource_type, root_scope, and routing_scope` => (resource_type, root_scope, routing_scope, created_at)

What do you think?

Reply:
Lemme run this through an explain plan and I'll share the results. There's a cost to extra indexes because they have to updated on writes. This should be easy to figure out with some data.

Reply:
@lakshmimsft and I did an experiment here 😁 

 I wanted to make sure that both forms of query (with routing scope and without) can hit the index. If they can't then that's a good reason to add a second index like you proposed.

First, I generated 20k rows of synthetic data - 10k for a top-level resource and 10k for a child resource.

## Generating 10k top-level resources

```sql
INSERT INTO resources (id, original_id, resource_type, root_scope, routing_scope, etag, resource_data)
SELECT 
  '/planes/radius/local/resourcegroups/default/providers/applications.core/containers/c' || numbers.id || '/' AS id,
  '/planes/radius/local/resourceGroups/default/providers/Applications.Core/containers/c' || numbers.id AS original_id,
  '/applications.core/containers/' as resource_type,
  '/planes/radius/local/resourcegroups/default/' as root_scope,
  '/applications.core/containers/c' || numbers.id || '/' as routing_scope,
  'etag' || numbers.id  as etag,
  '{}' as resource_data
FROM generate_series(1, 10000) AS numbers (id);
```


## Generating 10k child resources

```sql
INSERT INTO resources (id, original_id, resource_type, root_scope, routing_scope, etag, resource_data)
SELECT 
  '/planes/radius/local/resourcegroups/default/providers/applications.core/containers/c/' || numbers.id || '/identities/a/' AS id,
  '/planes/radius/local/resourceGroups/default/providers/Applications.Core/containers/c' || numbers.id || '/identities/a' AS original_id,
  '/applications.core/containers/identities/' as resource_type,
  '/planes/radius/local/resourcegroups/default/' as root_scope,
  '/applications.core/containers/c' || numbers.id || '/identities/a/' as routing_scope,
  'etag' || numbers.id  as etag,
  '{}' as resource_data
FROM generate_series(1, 10000) AS numbers (id);
```

Next, I took our client code and translated it into some candidate queries. This is what we're going to explain. I tested each query to make sure it was returning results, before running the query through explain.

These queries correspond with the resource provider "list" operations.

## Query top-level resources

```sql
SELECT original_id, etag, resource_data, created_at 
FROM resources
WHERE ((root_scope = '/planes/radius/local/resourcegroups/default/') OR (FALSE AND (root_scope LIKE '/planes/radius/local/resourcegroups/default/' || '%'))) AND 
	resource_type = '/applications.core/containers/' AND 
	((routing_scope LIKE NULL || '%') OR NULL IS NULL) AND 
	(created_at > NULL::TIMESTAMP OR NULL IS NULL)
ORDER BY created_at ASC
LIMIT 10;

```

## Query child resources


```sql

SELECT original_id, etag, resource_data, created_at 
FROM resources
WHERE ((root_scope = '/planes/radius/local/resourcegroups/default/') OR (FALSE AND (root_scope LIKE '/planes/radius/local/resourcegroups/default/' || '%'))) AND 
	resource_type = '/applications.core/containers/identities/' AND 
	((routing_scope LIKE '/applications.core/containers/c10/identities/a/' || '%') OR '/applications.core/containers/c10/identities/a/' IS NULL) AND 
	(created_at > NULL::TIMESTAMP OR NULL IS NULL)
ORDER BY created_at ASC
LIMIT 10;
```

## Findings

First of all, 20k rows wasn't enough data. At 20k rows PostreSQL did table scans and ignored the indices. Next we tried going to 2m rows.

Secondly, our current index has fields that are used in `ORDER BY` or with the `LIKE` operator. We get better results by taking out `routing_scope` and `created_at`.

```sql
CREATE INDEX idx_resource_query ON resources (resource_type, root_scope, routing_scope, created_at);
```

The explain-plan for a top-level resource query:

```txt
 Limit  (cost=153798.23..153799.40 rows=10 width=113)
   ->  Gather Merge  (cost=153798.23..250907.38 rows=832306 width=113)
         Workers Planned: 2
         ->  Sort  (cost=152798.21..153838.59 rows=416153 width=113)
               Sort Key: created_at
               ->  Parallel Seq Scan on resources  (cost=0.00..143805.29 rows=416153 width=113)
                     Filter: ((root_scope = '/planes/radius/local/resourcegroups/default/'::text) AND (resource_type = '/applications.core/containers/'::text))
```

Notice that the index is not used. This is the primary operation we should be concerned about having good performance.

----

The explain-plan for a child resource query:

```txt
 Limit  (cost=6.46..6.46 rows=1 width=113)
   ->  Sort  (cost=6.46..6.46 rows=1 width=113)
         Sort Key: created_at
         ->  Index Scan using idx_resource_query on resources  (cost=0.68..6.45 rows=1 width=113)
               Index Cond: ((resource_type = '/applications.core/containers/identities/'::text) AND (root_scope = '/planes/radius/local/resourcegroups/default/'::text))
               Filter: (routing_scope ~~ '/applications.core/containers/c10/identities/a/%'::text)
```

The operation is much less common and will need to handle much less data. The index is used, but `created_at` is sorted outside the index. `routing_scope` is used from the index, but it's a filter rather than a lookup due to the like.

## Better option

Here's what I decided to go with:

```sql
CREATE INDEX idx_resource_query ON resources (resource_type, root_scope);
```